doc/manual/src/architecture/architecture.md

```diff
@@ -0,0 +1,79 @@
+# Architecture
+
+*(This chapter is unstable and a work in progress. Incoming links may rot.)*
+
+This chapter describes how Nix works.
+It should help users understand why Nix behaves as it does, and it should help developers understand how to modify Nix and how to write similar tools.
+
+## Overview
+
+Nix consists of [hierarchical layers][layer-architecture].
+
+```
++-----------------------------------------------------------------+
+| Nix                                                             |
+|                  [ commmand line interface ]------,             |
+|                               |                   |             |
+|                           evaluates               |             |
+|                               |                manages          |
+|                               V                   |             |
+|                  [ configuration language  ]      |             |
+|                               |                   |             |
+| +-----------------------------|-------------------V-----------+ |
+| | store                  evaluates to                         | |
+| |                             |                               | |
+| |             referenced by   V       builds                  | |
+| |  [ build input ] ---> [ build plan ] ---> [ build result ]  | |
+| |                                                             | |
+| +-------------------------------------------------------------+ |
++-----------------------------------------------------------------+
+```
+
+At the top is the [command line interface](../command-ref/command-ref.md), translating from invocations of Nix executables to interactions with the underlying layers.
+
+Below that is the [Nix expression language](../expressions/expression-language.md), a [purely functional][purely-functional-programming] configuration language.
```

Comment:
When I started thinking about the term "Nix", this is the hierarchy that helped me put things into perspective:

```text

       ┌────────── Nix MODEL ──────────┐
       │                               │
       │ (This would be a spec, but    │
       │  it is the implementation     │
       │  itself, right?)              │
       │                               │
       └───────────────┬───────────────┘
                       │
                       │
                       │
                       ▼
┌───────────── Nix IMPLEMENTATION ──────────┐
│                 ("the core")              │
│                                           │    e.g., modules/c++ files
│         Implementation of the model       │          implementing flakes
│ (C++ code defining the building blocks of │
│  Nix: Nix lexer & parser, behaviour of    │
│  the Nix store, store objects and actions │
│  etc.)                                    │
│                                           │
└──────────────────────┬────────────────────┘
                       │
                       │
                       │
                       ▼
┌─────────────── Nix FRONTEND ──────────────┐            
│                                           │
│   + CLI tools that use the "core" libs    │    e.g., `nix flakes` commands
│                                           │
│   + Nix lang                              │
│     (Well, it is interpreted by the CLI   │
│      tools, but this is what is used to   │
│      interact with the system. Could be   │
│      wrong.)                              │
│                                           │
└───────────────────────────────────────────┘

```